sentry: conn_executor.go:643: panic while executing 1 statements: SELECT DISTINCT _.* FROM _ AS _ JOIN _ AS _ ON ((_._ = _._) AND (_._ = _._)) AND (_._ IS _) WHERE _._ = $1: caused by <redacted>

This issue was autofiled by Sentry. It represents a crash or reported error on a live cluster with telemetry enabled.

Sentry link: https://sentry.io/cockroach-labs/cockroachdb/issues/820548164/?referrer=webhooks_plugin

Panic message:

>conn_executor.go:643: panic while executing 1 statements: SELECT DISTINCT _.* FROM _ AS _ JOIN _ AS _ ON ((_._ = _._) AND (_._ = _._)) AND (_._ IS _) WHERE _._ = $1: caused by <redacted>


<details>
<summary>Stacktrace (expand for inline code snippets):</summary>

https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/conn_executor.go#L386-L388 in pkg/sql.(*Server).ServeConn.func1
/usr/local/go/src/runtime/asm_amd64.s#L572-L574 in runtime.call32
/usr/local/go/src/runtime/panic.go#L501-L503 in runtime.gopanic
/usr/local/go/src/runtime/panic.go#L27-L29 in runtime.panicindex
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/distsql_physical_planner.go#L934-L936 in pkg/sql.(*DistSQLPlanner).convertOrdering
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/distsql_physical_planner.go#L1917-L1919 in pkg/sql.(*DistSQLPlanner).createPlanForLookupJoin
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/distsql_physical_planner.go#L2272-L2274 in pkg/sql.(*DistSQLPlanner).createPlanForNode
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/distsql_physical_planner.go#L2278-L2280 in pkg/sql.(*DistSQLPlanner).createPlanForNode
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/distsql_physical_planner.go#L2657-L2659 in pkg/sql.(*DistSQLPlanner).createPlanForDistinct
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/distsql_physical_planner.go#L2328-L2330 in pkg/sql.(*DistSQLPlanner).createPlanForNode
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/distsql_physical_planner.go#L2278-L2280 in pkg/sql.(*DistSQLPlanner).createPlanForNode
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/distsql_running.go#L751-L753 in pkg/sql.(*DistSQLPlanner).PlanAndRun
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/conn_executor_exec.go#L981-L983 in pkg/sql.(*connExecutor).execWithDistSQLEngine
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/conn_executor_exec.go#L823-L825 in pkg/sql.(*connExecutor).dispatchToExecutionEngine
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/conn_executor_exec.go#L401-L403 in pkg/sql.(*connExecutor).execStmtInOpenState
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/conn_executor_exec.go#L95-L97 in pkg/sql.(*connExecutor).execStmt
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/conn_executor.go#L1172-L1174 in pkg/sql.(*connExecutor).run
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/conn_executor.go#L388-L390 in pkg/sql.(*Server).ServeConn
https://github.com/cockroachdb/cockroach/blob/0c87b11cb99ba5c677c95ded55dcba385928474e/pkg/sql/pgwire/conn.go#L312-L314 in pkg/sql/pgwire.(*conn).serveImpl.func4
</details>

```
pkg/sql/conn_executor.go in pkg/sql.(*Server).ServeConn.func1 at line 387
/usr/local/go/src/runtime/asm_amd64.s in runtime.call32 at line 573
/usr/local/go/src/runtime/panic.go in runtime.gopanic at line 502
/usr/local/go/src/runtime/panic.go in runtime.panicindex at line 28
pkg/sql/distsql_physical_planner.go in pkg/sql.(*DistSQLPlanner).convertOrdering at line 935
pkg/sql/distsql_physical_planner.go in pkg/sql.(*DistSQLPlanner).createPlanForLookupJoin at line 1918
pkg/sql/distsql_physical_planner.go in pkg/sql.(*DistSQLPlanner).createPlanForNode at line 2273
pkg/sql/distsql_physical_planner.go in pkg/sql.(*DistSQLPlanner).createPlanForNode at line 2279
pkg/sql/distsql_physical_planner.go in pkg/sql.(*DistSQLPlanner).createPlanForDistinct at line 2658
pkg/sql/distsql_physical_planner.go in pkg/sql.(*DistSQLPlanner).createPlanForNode at line 2329
pkg/sql/distsql_physical_planner.go in pkg/sql.(*DistSQLPlanner).createPlanForNode at line 2279
pkg/sql/distsql_running.go in pkg/sql.(*DistSQLPlanner).PlanAndRun at line 752
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).execWithDistSQLEngine at line 982
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).dispatchToExecutionEngine at line 824
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).execStmtInOpenState at line 402
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).execStmt at line 96
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).run at line 1173
pkg/sql/conn_executor.go in pkg/sql.(*Server).ServeConn at line 389
pkg/sql/pgwire/conn.go in pkg/sql/pgwire.(*conn).serveImpl.func4 at line 313
```

| Tag | Value |
|---|---|
| Cockroach Release | `v2.1.3` |
| Cockroach SHA: | 0c87b11cb99ba5c677c95ded55dcba385928474e |
| Platform | linux amd64 |
| Distribution | CCL |
| Environment | v2.1.3 |
| Command | server |
| Go Version | `go1.10.3`|
| # of CPUs |2 |
| # of Goroutines |145 |

I created a tiny repo that you can use to test: https://github.com/nexdrew/cockroach-33342-repro

It reliably produces this error for me:

```
E190104 18:15:22.268081 1288 sql/conn_executor.go:636  [n1,client=127.0.0.1:41052,user=root] a SQL panic has occurred while executing "SELECT q.comp_plan_id, count(q.id) AS count FROM test.public.quota AS q INNER JOIN test.public.comp_plan AS p ON p.id = q.comp_plan_id WHERE ((p.org_id = 'cjo63dwkz0000p8j7hytgl4jx') AND (p.deleted = false)) AND (q.deleted = false) GROUP BY q.comp_plan_id": runtime error: index out of range
E190104 18:15:22.268598 1288 util/log/crash_reporting.go:203  [n1,client=127.0.0.1:41052,user=root] a panic has occurred!
E190104 18:15:22.662802 1288 util/log/crash_reporting.go:477  [n1,client=127.0.0.1:41052,user=root] Reported as error 89796b1532cc48e8a8e97da6bf7c22c5
panic: runtime error: index out of range [recovered]
	panic: panic while executing 1 statements: SELECT _._, count(_._) AS _ FROM _._._ AS _ INNER JOIN _._._ AS _ ON _._ = _._ WHERE ((_._ = _) AND (_._ = _)) AND (_._ = _) GROUP BY _._; caused by runtime error: index out of range

goroutine 1288 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc42c9cf500, 0x3032640, 0xc42d35f400, 0x2770200, 0x4231d70)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:650 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc42c9cf500, 0x3032640, 0xc42d35f400)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:387 +0x61
panic(0x2770200, 0x4231d70)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).convertOrdering(0xc420903600, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:935 +0x27a
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPlanForLookupJoin(0xc420903600, 0xc42d64a660, 0xc42d6484e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1918 +0x964
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPlanForNode(0xc420903600, 0xc42d64a660, 0x3025380, 0xc42d6484e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:2273 +0x1284
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPlanForLookupJoin(0xc420903600, 0xc42d64a660, 0xc42d648680, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:1844 +0xba
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPlanForNode(0xc420903600, 0xc42d64a660, 0x3025380, 0xc42d648680, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:2273 +0x1284
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPlanForNode(0xc420903600, 0xc42d64a660, 0x3025700, 0xc42d623600, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:2279 +0xccd
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).createPlanForNode(0xc420903600, 0xc42d64a660, 0x3025180, 0xc42d655200, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:2289 +0x863
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc420903600, 0x3032700, 0xc42d572d50, 0xc42c9cf9b0, 0xc42d64a660, 0xc42cb17440, 0x3025180, 0xc42d655200, 0xc42d65ca00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:752 +0x126
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc42c9cf500, 0x3032700, 0xc42d572d50, 0xc42c9cf918, 0x3, 0x7f52136cb950, 0xc42cb174d0, 0x1, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:982 +0x2d8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc42c9cf500, 0x3032700, 0xc42d572d50, 0x3035d80, 0xc42cbc3c00, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:824 +0xa8a
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc42c9cf500, 0x3032700, 0xc42d572d50, 0x3035d80, 0xc42cbc3c00, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:402 +0xb3b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc42c9cf500, 0x3032700, 0xc42d572d50, 0x3035d80, 0xc42cbc3c00, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:96 +0x341
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc42c9cf500, 0x3032640, 0xc42d35f400, 0xc4209af098, 0x5400, 0x15000, 0xc4209af130, 0xc4297cd490, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1116 +0x21aa
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc4209ba960, 0x3032640, 0xc42d35f400, 0xc42c9cf500, 0x5400, 0x15000, 0xc4209af130, 0xc4297cd490, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:389 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4(0xc4209ba960, 0x3032640, 0xc42d35f400, 0xc42c9cf500, 0x5400, 0x15000, 0xc4209af130, 0xc4297cd490, 0xc4297cd4a0, 0xc42cea7e60)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:313 +0x81
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:312 +0x107c
```

Hopefully this helps!

--- a/pkg/sql/distsql_physical_planner.go
+++ b/pkg/sql/distsql_physical_planner.go
@@ -1915,7 +1915,7 @@ func (dsp *DistSQLPlanner) createPlanForLookupJoin(
 		distsqlrun.ProcessorCoreUnion{JoinReader: &joinReaderSpec},
 		post,
 		types,
-		dsp.convertOrdering(planPhysicalProps(n), plan.PlanToStreamColMap),
+		dsp.convertOrdering(planPhysicalProps(n), planToStreamColMap),
 	)
 	plan.PlanToStreamColMap = planToStreamColMap
 	return plan, nil

--- a/pkg/sql/logictest/testdata/logic_test/lookup_join
+++ b/pkg/sql/logictest/testdata/logic_test/lookup_join
@@ -498,3 +498,15 @@ query I
 SELECT u.a FROM t JOIN u ON t.d = u.d AND t.a = u.a AND t.d = u.d WHERE t.e = 5
 ----
 1
+
+# Regression test for #33342. We want the ordering in the lookup join node to
+# be +f not +a; this is achieved by making def show up in the query first (and
+# thus get smaller opt.ColumnIDs).
+query IIIIII colnames,partialsort(4)
+SELECT * FROM def JOIN abc ON a=f ORDER BY a
+----
+d     e  f  a  b     c
+2     1  1  1  1     2
+NULL  2  1  1  1     2
+1     1  2  2  NULL  2
+1     1  2  2  1     1

--- a/pkg/sql/opt/exec/execbuilder/testdata/lookup_join
+++ b/pkg/sql/opt/exec/execbuilder/testdata/lookup_join
@@ -560,3 +560,24 @@ render            ·         ·          (a)              ·
       │           filter    e = 5      ·                ·
       └── scan    ·         ·          (a, d)           ·
 ·                 table     u@idx      ·                ·
+
+# Plan for #33342 regression test. We want the ordering in the lookup join node
+# to be `+f` not `+a`; this is achieved by making def show up in the query
+# first (and thus get smaller opt.ColumnIDs).
+query TTTTT
+EXPLAIN (VERBOSE) SELECT * FROM def JOIN abc ON a=f ORDER BY a
+----
+render            ·         ·            (d, e, f, a, b, c)  ·
+ │                render 0  d            ·                   ·
+ │                render 1  e            ·                   ·
+ │                render 2  f            ·                   ·
+ │                render 3  a            ·                   ·
+ │                render 4  b            ·                   ·
+ │                render 5  c            ·                   ·
+ └── lookup-join  ·         ·            (a, b, c, d, e, f)  +f
+      │           type      inner        ·                   ·
+      ├── scan    ·         ·            (a, b, c)           +a
+      │           table     abc@primary  ·                   ·
+      │           spans     ALL          ·                   ·
+      └── scan    ·         ·            (d, e, f)           ·
+·                 table     def@primary  ·                   ·

